gnomad_hail/resources/resource_utils.py

```diff
@@ -9,114 +9,221 @@
+            import_func: Optional[Callable] = None,
+            expected_file_extensions: Optional[List[str]] = None
+    ):
+        if path is None and import_func is None:
+            raise ValueError(f"{self.__class__.__name__} requires at least one of path or import_func arguments.")
 
         self.path = path
         self.import_sources = import_sources
+        self.import_func = import_func
 
-        if expected_file_extension and not path.endswith(expected_file_extension):
+        if path is not None and expected_file_extensions and not [ext for ext in expected_file_extensions if path.endswith(ext)]:
             logger.warning(
-                f"Created the following {self.__class__.__name__} with a path that doesn't ends with {expected_file_extension}: {self}")
+                "Created the following {} with a path that doesn't ends with {}: {}".format(
+                    self.__class__.__name__,
+                    " or ".join(expected_file_extensions),
+                    self
+                )
+            )
 
     def __repr__(self):
         attr_str = [f'path={self.path}']
         if self.import_sources is not None:
             attr_str.append(f'import_sources={self.import_sources}')
         return f'{self.__class__.__name__}({",".join(attr_str)})'
 
+    @abstractmethod
+    def import_resource(self):
+        """
+        Abstract method to import the resource using its import_func and writes it in its path.
+        """
+        pass
+
 
 class TableResource(BaseResource):
     """
     A Hail Table resource
+
+    :param path: The Table path (typically ending in .ht)
+    :param import_sources: Any sources that are required for the import and need to be kept track of and/or passed to the import_func (e.g. .vcf path for an imported VCF)
+    :param import_func: A function used to import the Table. `import_func` will be passed the `import_sources` dictionary as kwargs.
     """
 
-    def __init__(self, path: str, import_sources: Optional[Dict[str, Any]] = None, ):
+    def __init__(
+            self,
+            path: Optional[str] = None,
+            import_sources: Optional[Dict[str, Any]] = None,
+            import_func: Optional[Callable[..., hl.Table]] = None
+    ):
         super().__init__(
             path=path,
             import_sources=import_sources,
-            expected_file_extension='.ht'
+            import_func=import_func,
+            expected_file_extensions=['.ht']
         )
 
-    def ht(self) -> hl.Table:
+    def ht(self, force_import: bool = False) -> hl.Table:
         """
         Read and return the Hail Table resource
 
         :return: Hail Table resource
         """
-        return hl.read_table(self.path)
+        if self.path is None or force_import:
+            return self.import_func(**self.import_sources)
+        else:
+            return hl.read_table(self.path)
+
+    def import_resource(self, overwrite: bool = False, **kwargs):
```

Comment:
If arbitrary kwargs can be passed to `Table.write`, is there any reason to separate `overwrite` from other kwargs? Same question for MatrixTableResource.

Reply:
yes, `overwrirte` is used all the time, the other args essentially never. I still thought that adding this pass-over functionality could potentially be helpful.